### ISSUE TITLE
"Function" variables in Optimizer

### ISSUE THREAD

**Issue by PaulTalbot-INL** [idaholab/raven#867]
--------
Issue Description
--------
In other Samplers, a `<variable>` can be entirely redundant (dependent on another sampled `<variable>`) by providing a `<function>` instead of a `<distribution>`.

This functionality would be useful in the Optimizer, but currently the Optimizer reads all `<variable>` blocks without the possibility of providing a `<function>` instead of the optimization bounds.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

**Comment by PaulTalbot-INL** [comment]
@elementx54 This reflects the conversation we had earlier today.

**Comment by elementx54** [comment]
Yes. That is exactly what I was talking about. A sampler can let another variable depend on the value of a varying variable, but the optimizer has no way to specify this relationship. As the sampler handles providing values to variables, I was expecting the optimizer to fulfill the same role with all my variables (constant, independent varying, and dependent varying) in a similar way.

The work around of using an `ExternalModel` within an `EnsembleModel` to create the value of the dependent variables from the varying independent variables is much more work and prone to error and confusion.

**Comment by alfoa** [comment]
@PaulTalbot-INL I honestly did not get where the problem is.

Can you explain it a bit better ?

Thanks

**Comment by PaulTalbot-INL** [comment]
Ways to have a variable value come from a Sampler (all examples within a Sampler block):
 - Getting Sampled from a Distribution
```
  <variable name='x'>
    <distribution>x_dist</distribution>
  </variable>
```
 - Getting co-sampled from a Distribution
```
  <variable name='x,y'>
    <distribution>x_dist</distribution>
  </variable>
```
 - As a Constant
```
  <constant name='x'>3</constant>
```
 - As a Function (aka redundant evaluation)
```
  <variable name='x'>
    <function>getRadiusFromCircumference</function>
  </variable>
```
The last entry is NOT available in the Optimizer.

**Comment by wangcj05** [comment]
This issue can be closed with PR #1163
